tempodb/encoding/versioned.go

```diff
@@ -15,15 +15,26 @@ import (
 type VersionedEncoding interface {
 	Version() string
 
-	NewDataWriter(writer io.Writer, encoding backend.Encoding) (common.DataWriter, error)
-	NewIndexWriter(pageSizeBytes int) common.IndexWriter
+	// OpenBlock for reading
+	OpenBlock(meta *backend.BlockMeta, r backend.Reader) (common.BackendBlock, error)
 
-	NewDataReader(ra backend.ContextReader, encoding backend.Encoding) (common.DataReader, error)
-	NewIndexReader(ra backend.ContextReader, pageSizeBytes int, totalPages int) (common.IndexReader, error)
+	// NewCompactor creates a Compactor that can be used to combine blocks of this
+	// encoding. It is expected to use internal details for efficiency.
+	NewCompactor() common.Compactor
```

Comment:
I lean towards option 1, to only compact blocks of the same encoding, as it is the most performant since encodings can make use of internal details.  Added a note to our implementation checklist.

Also considered option 2, use TraceIterator, because there are similarities between WAL -> complete block and compaction (both are creating a new block with known trace contents).  The flexibility is much better, and allows upgrading of existing blocks to new formats.  However, performance would be reduced for normal compaction which is the majority of work.

Reply:
+1 to option 1. Currently we don't compact across data encodings:

https://github.com/grafana/tempo/blob/main/tempodb/compaction_block_selector.go#L126 

As a concept "dataEncoding" goes away with parquet blocks. Perhaps it means something else? or we drop the field entirely?

Reply:
Oops we are mixing up terminology which is clouding the discussion.  The compactor needs to select blocks with the same `BlockMeta.Version` and `BlockMeta.DataEncoding`, i.e. v2->v2 and parquet->parquet.  `BlockMeta.Encoding` is actually the compression which doesn't matter.   Agree that the first parquet format won't make use of DataEncoding to start, so we can either leave it as an empty string for further use, or remove it.